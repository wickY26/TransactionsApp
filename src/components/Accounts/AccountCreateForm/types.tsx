import { Account } from '../../../containers/Accounts/types';

export interface AccountCreateFormProps {
  onCreateAccount: (data: Account) => void;
}
