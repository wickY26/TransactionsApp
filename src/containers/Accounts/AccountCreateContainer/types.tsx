import { ReactElement } from 'react';

import { Account } from '../types';

export interface RenderProps {
  onCreateAccount: (data: Account) => void;
}

export interface AccountCreateContainerProps {
  children: (props: RenderProps) => ReactElement;
};
