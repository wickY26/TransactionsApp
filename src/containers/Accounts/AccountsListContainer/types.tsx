import { ReactElement } from 'react';

import { Account } from '../types';

export interface RenderProps {
  accounts: Account[];
}

export interface AccountsListContainerProps {
  children: (props: RenderProps) => ReactElement;
};
