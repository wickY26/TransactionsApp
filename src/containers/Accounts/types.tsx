import { Action } from 'redux';

import { ActionTypes as at } from './constant';

export interface Account {
  name: string;
  surname: string;
  id?: number;
};

export interface FetchAccountsSuccessAction extends Action {
  type: at.FetchAccountsSuccess;
  payload: Account[];
}

export interface FetchAccountsErrorAction extends Action {
  type: at.FetchAccountsError;
}

export interface CreateAccountSuccessAction extends Action {
  type: at.CreateAccountSuccess;
  payload: Account;
}

export interface CreateAccountErrorAction extends Action {
  type: at.CreateAccountError;
}

export interface State {
  accounts: Account[];
}

export type AccountsActions = 
  | FetchAccountsSuccessAction
  | FetchAccountsErrorAction
  | CreateAccountSuccessAction
  | CreateAccountErrorAction
