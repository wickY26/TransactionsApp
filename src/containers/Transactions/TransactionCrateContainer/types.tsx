import { ReactElement } from 'react';

import { TransactionBody } from '../types';

export interface RenderProps {
  onCreateTransaction: (data: TransactionBody) => void;
}

export interface TransactionCreateContainerProps {
  children: (props: RenderProps) => ReactElement;
};
