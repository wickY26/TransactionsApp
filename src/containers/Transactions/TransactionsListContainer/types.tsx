import { ReactElement } from 'react';

import { Transaction } from '../types';

export interface RenderProps {
  transactions: Transaction[];
}

export interface TransactionsListContainerProps {
  children: (props: RenderProps) => ReactElement;
};
