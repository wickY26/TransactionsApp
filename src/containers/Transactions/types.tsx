import { Action } from 'redux';

import { ActionTypes as at } from './constant';

export interface TransactionAccount {
  name: string;
  surname: string;
  id: number;
}

export interface Transaction {
  from: TransactionAccount;
  to: TransactionAccount;
  amount: number;
  description: string;
  id: number;
};

export interface TransactionBody {
  from: number;
  to: number;
  amount: number;
  description: string;
};

export interface FetchTransactionsSuccessAction extends Action {
  type: at.FetchTransactionsSuccess;
  payload: Transaction[];
}

export interface FetchTransactionsErrorAction extends Action {
  type: at.FetchTransactionsError;
}

export interface CreateTransactionSuccessAction extends Action {
  type: at.CreateTransactionSuccess;
  payload: Transaction;
}

export interface CreateTransactionErrorAction extends Action {
  type: at.CreateTransactionError;
}

export interface State {
  transactions: Transaction[];
}

export type TransactionsActions = 
  | FetchTransactionsSuccessAction
  | FetchTransactionsErrorAction
  | CreateTransactionSuccessAction
  | CreateTransactionErrorAction
