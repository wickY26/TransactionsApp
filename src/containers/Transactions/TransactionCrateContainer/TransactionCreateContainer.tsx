import React, { FC, useState } from 'react';
import { Redirect } from 'react-router-dom';
import { useDispatch } from 'react-redux';

import { TransactionCreateContainerProps } from './types';
import { thunkCreateTransaction } from '../actions';
import { TransactionBody } from '../types';

const TransactionCreateContainer: FC<TransactionCreateContainerProps> = ({ children }) => {
  const [redirect, setRedirect] = useState<boolean>(false);
  const dispatch = useDispatch();

  const handleOnCreateTransaction = (transaction: TransactionBody) => {
    dispatch(thunkCreateTransaction(transaction));
    setRedirect(true);
  }

  return (
    <>
      {redirect && <Redirect push to="/transactions" />}
      {children({ onCreateTransaction: handleOnCreateTransaction })}
    </>
  );
};

export default TransactionCreateContainer;
