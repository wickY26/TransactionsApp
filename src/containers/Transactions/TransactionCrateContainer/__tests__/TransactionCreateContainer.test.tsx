import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import TransactionCreateContainer from '..';
import { TransactionBody } from '../../types';

const mockTransactionBody: TransactionBody = { from: 0, to: 1, amount: 50, description: 'description' };
const mockDispatch = jest.fn();
const mockThunkCreateTransaction = jest.fn();

jest.mock('react-redux', () => ({
  useDispatch: () => mockDispatch,
}));

jest.mock('react-router-dom', () => ({
  Redirect: () => 'Redirect',
}));

jest.mock('../../actions', () => ({
  thunkCreateTransaction: (transaction: any) => mockThunkCreateTransaction(transaction),
}));

const renderComponent = () => render(
  <TransactionCreateContainer>
    {({onCreateTransaction}) => (
      <button onClick={()=> onCreateTransaction(mockTransactionBody)}>Button</button>
    )}
  </TransactionCreateContainer>
);

describe('TransactionCreateContainer', () => {

  beforeEach(() => {
    mockThunkCreateTransaction.mockReturnValue('CreateUserAction');
  });

  it('should render correctly', () => {
    const component = renderComponent();
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should render redirect when onCreateTransaction is called', () => {
    const component = renderComponent();
    fireEvent.click(screen.getByText('Button'));
    expect(component.asFragment()).toMatchSnapshot();
  });
  
  it('should dispatch createTransactionAction when onCreateTransaction is called', () => {
    renderComponent();
    fireEvent.click(screen.getByText('Button'));
    expect(mockThunkCreateTransaction).toHaveBeenCalledWith(mockTransactionBody);
    expect(mockDispatch).toHaveBeenCalledWith('CreateUserAction');
  });

});
