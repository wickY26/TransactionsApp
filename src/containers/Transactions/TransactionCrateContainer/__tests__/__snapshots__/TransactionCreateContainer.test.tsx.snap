// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionCreateContainer should render correctly 1`] = `
<DocumentFragment>
  <button>
    Button
  </button>
</DocumentFragment>
`;

exports[`TransactionCreateContainer should render redirect when onCreateTransaction is called 1`] = `
<DocumentFragment>
  Redirect
  <button>
    Button
  </button>
</DocumentFragment>
`;
